proto-grpc.md

```diff
@@ -0,0 +1,254 @@
+  of files and targets without having to read the contents of any of the source files. Resolving
+  this behaviour into a structure that can work under Bazel leads to a large part of the variation
+  between rulesets. At a basic level, a number of these plugins can be manipulated into a
+  deterministic behaviour through options and post-processing, or the use of
+  [declare_directory](https://docs.bazel.build/versions/main/skylark/lib/actions.html#declare_directory)
+  to capture the full output tree without knowing its structure ahead of time.
+
+  However, some plugins are sufficiently odd that the above process is unsuitable or unable to
+  capture the behaviour; this is where [stackb/rules_proto](https://github.com/stackb/rules_proto)
+  introduces a [Gazelle](https://github.com/bazelbuild/bazel-gazelle) plugin, which directly parses
+  the `.proto` files, captures any of the plugin behaviour in Go functions and produces `BUILD`
+  files that satisfy all of Bazel's requirements.
+
+- **Cross-Cutting**: Rulesets that support Protobuf and gRPC need to know about both protoc
+  execution and about the target languages for which the plugins are generating files. This requires
+  the authors of such rulesets to have expertise in every language that they hope to support, whilst
+  also having expertise in how protoc itself works, which narrows the pool of people interested in
+  tackling this.
+
+- **Dependencies**: Due to the above cross-cutting nature, a ruleset that supports many languages
+  also ends up with many dependencies, on both other rulesets and on Protobuf/gRPC implementations.
+  Due to how dependencies currently work with Bazel workspaces, this typically results in users
+  stuck with a Protobuf/gRPC version dictated by their choice of rulesets, unless they are
+  comfortable overriding dependencies. This also makes maintenance of these rulesets more
+  challenging, since rules authors must keep on top of updates to their upstream dependencies.
+
+  The introduction of [bzlmod](https://docs.bazel.build/versions/main/bzlmod.html) to Bazel 
+  (expected in version 6) will hopefully resolve many of the dependency issues listed above and in
+  doing so remove workload from rules authors and allow rules users more freedom to choose versions.
+  Under bzlmod, the dependency tree is explicitly managed, with updates to transitive dependencies
+  handled transparently without updates to intermidiary rulesets.
+
+The above three points dictate much of fragmentation seen between rulesets, since there are numerous
+solutions to these problems, each with their own tradeoffs. The solution that works well for one
+language may be less suitable for another, leading to naturally diverging behaviour. As such,
+hopefully it is now slightly cleared why there is such a diversity of Protobuf and gRPC rulesets,
+with no single 'right answer' that solves everyones' problems.
+
+
+## Custom Plugins
```

Comment:
@Yannic has mentioned that the choice of language for the plugin matters a lot too - if you can write them in C++ you avoid more dependency hassle

Reply:
Is there a one-line summary for why this is the case, which I can add?

Reply:
Writing a well-behaved protoc plugin is a topic on its own and can't be compressed to a single sentence. It's not as much the language that matters and more that the plugin follows the conventions mentioned above, so we cover some of it. A "how to write a well-behaved protoc plugin" seems out of scope for this.

Reply:
Ok, sounds like we're good as is then. The mention of Go in this section is about rules_proto's Gazelle plugin, rather than for a protoc plugin